crates/engine/src/test_api.rs

```diff
@@ -0,0 +1,322 @@
+///
+/// # Note
+///
+/// Note that account could refer to either a user account or
+/// a smart contract account. This returns the same as `env::api::balance`
+/// if given the account ID of the currently executed smart contract.
+///
+/// # Errors
+///
+/// - If `account` does not exist.
+/// - If the underlying `account` type does not match.
+pub fn get_account_balance<T>(_account_id: T::AccountId) -> Result<T::Balance>
+where
+    T: Environment,
+{
+    unimplemented!("off-chain environment does not yet support `get_account_balance`");
+}
+
+/// Sets the rent allowance of the contract account to the given rent allowance.
+///
+/// # Errors
+///
+/// - If `account` does not exist.
+/// - If the underlying `account` type does not match.
+/// - If the underlying `new_rent_allowance` type does not match.
+pub fn set_contract_rent_allowance<T>(
+    _account_id: T::AccountId,
+    _new_rent_allowance: T::Balance,
+) -> Result<()>
+where
+    T: Environment,
+{
+    unimplemented!(
+        "off-chain environment does not yet support `set_contract_rent_allowance`"
+    );
+}
+
+/// Returns the rent allowance of the contract account.
+///
+/// # Errors
+///
+/// - If `account` does not exist.
+/// - If the underlying `account` type does not match.
+/// - If the returned rent allowance cannot be properly decoded.
+pub fn get_contract_rent_allowance<T>(_account_id: T::AccountId) -> Result<T::Balance>
+where
+    T: Environment,
+{
+    unimplemented!(
+        "off-chain environment does not yet support `get_contract_rent_allowance`"
+    );
+}
+
+/// Set the entropy hash of the current block.
+///
+/// # Note
+///
+/// This allows to control what [`random`][`crate::random`] returns.
+pub fn set_block_entropy<T>(_entropy: T::Hash) -> Result<()>
+where
+    T: Environment,
+{
+    unimplemented!("off-chain environment does not yet support `set_block_entropy`");
+}
+
+/// Returns the contents of the past performed environmental `println` in order.
+pub fn recorded_printlns() -> impl Iterator<Item = String> {
+    // TODO
+    vec![String::from("")].into_iter()
+}
+
+/// Advances the chain by a single block.
+pub fn advance_block<T>() -> Result<()>
+where
+    T: Environment,
+{
+    unimplemented!("off-chain environment does not yet support `advance_block`");
+}
+
+/// Set to true to disable clearing storage
+///
+/// # Note
+///
+/// Useful for benchmarking because it ensures the initialized storage is maintained across runs,
+/// because lazy storage structures automatically clear their associated cells when they are dropped.
+pub fn set_clear_storage_disabled(_disable: bool) {
+    unimplemented!(
+        "off-chain environment does not yet support `set_clear_storage_disabled`"
+    );
+}
+
+/// Initializes the whole off-chain environment.
+///
+/// # Note
+///
+/// - Initializes the off-chain environment with default values that fit most
+/// uses cases.
+pub fn initialize_or_reset_as_default<T>() -> Result<()>
```

Comment:
It would be amazing if we did not need this. Maybe this is possible to achieve using `once_cell` or so. Not sure though. We might still be in need of a `reset` functionality though but manually initializing an environment before first use should definitely be automated. Maybe it already is that way but for ink!'s dynamic storage allocator it is a total mess.